trie/database.go

```diff
@@ -662,10 +662,12 @@ func (db *Database) Commit(node common.Hash, report bool, callback func(common.H
 	}
 	// Uncache any leftovers in the last batch
 	db.lock.Lock()
-	defer db.lock.Unlock()
-
-	batch.Replay(uncacher)
+	err := batch.Replay(uncacher)
 	batch.Reset()
+	db.lock.Unlock()
+	if err != nil {
+		return err
+	}
```

Comment:
I think this might be subtly wrong -- the lock previously also guarded e.g. `db.gcnodes`, which is also updated in `Dereference`. This change makes the lock scope smaller, no longer covering the updates on L:685-686

Reply:
Yes, in this case, we can just rely on the `defer` to do the `Unlock`.